### PR TITLE
[AIRFLOW-3960] Create Google Cloud Speech operators

### PR DESCRIPTION
**Pull request by antonimaciej**
### Jira

https://issues.apache.org/jira/browse/AIRFLOW-3960

### Description

Added two new operators:
GcpSpeechToTextRecognizeOperator
GcpTextToSpeechSynthesizeOperator

CC @kaxil 

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

TestTextToSpeechOperator, TestTextToSpeechHook, TestSpeechToTextOperator, TestSpeechToTextHook

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: airflow/contrib/hooks/gcp_speech_to_text_hook.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from google.cloud.speech_v1 import SpeechClient
+
+from airflow.contrib.hooks.gcp_api_base_hook import GoogleCloudBaseHook
+
+
+class GCPSpeechToTextHook(GoogleCloudBaseHook):
```
Comment (XD-DENG) [inline airflow/contrib/hooks/gcp_speech_to_text_hook.py:25]: Docstring is missing.

File: airflow/contrib/hooks/gcp_text_to_speech_hook.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from google.cloud.texttospeech_v1 import TextToSpeechClient
+
+from airflow.contrib.hooks.gcp_api_base_hook import GoogleCloudBaseHook
+
+
+class GCPTextToSpeechHook(GoogleCloudBaseHook):
```
Comment (XD-DENG) [inline airflow/contrib/hooks/gcp_text_to_speech_hook.py:25]: Docstring is missing.

File: setup.py [hunk 3]
```diff
@@ -180,6 +180,8 @@ def write_version(filename=os.path.join(*['airflow',
     'google-cloud-bigtable==0.31.0',
     'google-cloud-spanner>=1.7.1',
     'google-cloud-vision>=0.35.2',
+    'google-cloud-texttospeech',
+    'google-cloud-speech',
```
Comment (XD-DENG) [inline setup.py]: Better to specify version (range).

**Review by XD-DENG — CHANGES_REQUESTED** [review]
A few comments at first glance.

File: airflow/contrib/operators/gcp_text_to_speech_operator.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        if self.input_ == "":
+            raise AirflowException("The required parameter 'input_' is empty")
+        if self.voice == "":
+            raise AirflowException("The required parameter 'voice' is empty")
+        if self.audio_config == "":
+            raise AirflowException("The required parameter 'audio_config' is empty")
+        if self.target_bucket_name == "":
+            raise AirflowException("The required parameter 'target_bucket_name' is empty")
+        if self.target_filename == "":
+            raise AirflowException("The required parameter 'target_filename' is empty")
```
Comment (OmerJog) [inline airflow/contrib/operators/gcp_text_to_speech_operator.py]: Why not use pattern like:
```
for input in [self.project_id, self.input_, ...]:
    if getattr(self, input) == "":
        raise AirflowException(f"The required parameter '{input}' is empty")
```

File: airflow/contrib/hooks/gcp_speech_to_text_hook.py (hunk 1 above)
Comment (mik-laj) [reply airflow/contrib/hooks/gcp_speech_to_text_hook.py:25]: All hooks (except `GoogleCloudBaseHook`) for the GCP platform have very limited descriptions. What do you think should be included in this description? I am interested in the development of integration with the GCP platform and your opinion is valuable to me. What parts should the description consist of? 

https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_vision_hook.py#L87-L90
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_transfer_hook.py#L33-L37
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_spanner_hook.py#L27-L34
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_pubsub_hook.py#L40-L45
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_mlengine_hook.py#L48-L51
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_kms_hook.py#L38-L42
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_function_hook.py#L36-L42
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_function_hook.py#L36-L42
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_dataproc_hook.py#L206-L213
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_dataflow_hook.py#L176-L178
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_container_hook.py#L37-L42
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_compute_hook.py#L41-L46
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_bigtable_hook.py#L28-L34

Comment (XD-DENG) [reply airflow/contrib/hooks/gcp_speech_to_text_hook.py:25]: Hi @mik-laj , AWS-related hooks can be considered as reference, e.g. https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/aws_dynamodb_hook.py#L24-L34 .

The main missing piece to me here is the parameters and param types. Let's take https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_transfer_hook.py#L33-L37 as example: there are three parameters, `api_version`, `gcp_conn_id`, and `delegate_to`. For `api_version` & `delegate_to`: what's their usage/definition/purpose/valid values, as well as their types. These should be clear in the docstring IMO.

File: airflow/contrib/operators/gcp_text_to_speech_operator.py (hunk 4 above)
Comment (antonimaciej) [reply airflow/contrib/operators/gcp_text_to_speech_operator.py]: Good suggestion. Done

File: setup.py (hunk 3 above)
Comment (antonimaciej) [reply setup.py]: Sure, added that.

File: airflow/contrib/hooks/gcp_speech_to_text_hook.py (hunk 1 above)
Comment (antonimaciej) [reply airflow/contrib/hooks/gcp_speech_to_text_hook.py:25]: Params description added

File: airflow/contrib/hooks/gcp_text_to_speech_hook.py (hunk 2 above)
Comment (antonimaciej) [reply airflow/contrib/hooks/gcp_text_to_speech_hook.py:25]: Params description added

File: airflow/contrib/hooks/gcp_speech_to_text_hook.py (hunk 1 above)
Comment (mik-laj) [reply airflow/contrib/hooks/gcp_speech_to_text_hook.py:25]: Thanks for explaining. I will prepare a fix for all GCP hooks.

File: airflow/contrib/hooks/gcp_text_to_speech_hook.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        """
+        if not self._client:
+            self._client = TextToSpeechClient(credentials=self._get_credentials())
+        return self._client
+
+    def synthesize_speech(self, input_, voice, audio_config, retry=None, timeout=None):
+        """
+        Synthesizes text input
+
+        :param input_: text input to be synthesized. See more:
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_text_to_speech_hook.py]: really hate this kind of param names (ending in underscore) .. can we change `input_` to something else?

File: airflow/contrib/operators/gcp_speech_to_text_operator.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        self.config = config
+        self.project_id = project_id
+        self.gcp_conn_id = gcp_conn_id
+        self.retry = retry
+        self.timeout = timeout
+        self._validate_inputs()
+        super(GcpSpeechToTextRecognizeSpeechOperator, self).__init__(*args, **kwargs)
+
+    def _validate_inputs(self):
+        if self.project_id == "":
```
Comment (kaxil) [inline airflow/contrib/operators/gcp_speech_to_text_operator.py]: This shouldn't error, right? If project ID is not specified, it should be taken from GCP connection. There is a decorator @potiuk created that you can use.

**Review by kaxil — CHANGES_REQUESTED** [review]

File: airflow/contrib/hooks/gcp_text_to_speech_hook.py (hunk 5 above)
Comment (antonimaciej) [reply airflow/contrib/hooks/gcp_text_to_speech_hook.py]: Changed it to `input_data`

File: airflow/contrib/operators/gcp_speech_to_text_operator.py (hunk 6 above)
Comment (antonimaciej) [reply airflow/contrib/operators/gcp_speech_to_text_operator.py]: Added fallback_to_default_project_id decorator to hook's method, and removed project_id from validation

**Review by kaxil — APPROVED** [review]

**Comment by kaxil** [comment]
@XD-DENG Any more comments on this?

**Review by XD-DENG — APPROVED** [review]

**Comment by XD-DENG** [comment]
LGTM @kaxil @Fokko . (On a vacation so responding slowly, sorry for this)

**Review by mik-laj — APPROVED** [review]

**Pull request merged**